File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a, b) => a - b);
    const result = [];

    for (let i = 0; i < nums.length; i++) {
        if (i !== 0 && nums[i] === nums[i - 1]) continue;

        let left = i + 1,
          right = nums.length - 1;

        while (left < right) {
          if (nums[i] + nums[left] + nums[right] === 0) {
            result.push([nums[i], nums[left], nums[right]]);
            left++;
            while (nums[left] === nums[left - 1] && left < right) {
              left++;
            }
          } else if (nums[i] + nums[left] + nums[right] > 0) {
            right--;
          } else {
            left++;
          }
        }
    }

    return result;
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        
        for i, v in enumerate(nums):
            if i > 0 and v == nums[i - 1]:
                continue
            l = i + 1
            r = len(nums)  - 1
            while l < r:
                triplet = v + nums[l] + nums[r]
                if triplet == 0:
                    res.append([v, nums[l], nums[r]])
                    l += 1
                    while nums[l] == nums[l - 1] and l < r:
                        l += 1
                elif triplet > 0:
                    r -= 1
                else:
                    l += 1
                    
        return res
                